src/components/tagging/TagSelector.js

```diff
@@ -1,17 +1,17 @@
 // @flow
 import * as React from 'react';
 import shortid from 'shortid';
-import { connect } from 'react-redux';
-import ColorFactory from '../../factories/ColorFactory';
-import { addTag, removeLastTag } from '../../state/image/image.actionCreators';
-import type { Image } from '../../models/Image';
-import type { AppState } from '../../state/AppState';
+import Tag from '../utility/Tag';
+import type { Tag as TagType } from '../../models/Tag';
 import './TagSelector.scss';
 
-
 type Props = {
-  image: ?Image,
-  dispatch: Function,
+  tags: Array<TagType>,
+  multiTags: Array<TagType>,
```

Comment:
Is there a specific reason you chose to maintain two arrays instead of adding a boolean `isMulti` to the Tag-Type?
Also couldn't the tags be saved inside the image prop? So that you wrap everything inside image insttead of maintaining it in the Props?

Looking through this PR there seems to be a lot of duplication. I don't know much about react but I'm pretty close to calling it a shotgun surgery are you sure there is nothing that can be done about that with some design changes?

Reply:
Simply ctrl+f for multiTags brings up 15 results. Sure tests are included here but that still seems a bit excessive.

Reply:
And regardning the occurences, the `multiTags` Array is actually pretty lightweight. The array is stored in the `UploadContext` (which sadly forces many occurences in Context definition and the provider, [more about react context here](https://reactjs.org/docs/context.html#reactcreatecontext)) and the `TaggingForm` and `TaggingSelector` which need to access the multiTags to provide the highlighting.